internal/resources/metal/vlan/models.go

```diff
@@ -0,0 +1,98 @@
+package vlan
+
+import (
+	"context"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/packethost/packngo"
+	"strings"
+)
+
+type DataSourceModel struct {
+	ID                 types.String `tfsdk:"id"`
+	ProjectID          types.String `tfsdk:"project_id"`
+	VlanID             types.String `tfsdk:"vlan_id"`
+	Vxlan              types.Int64  `tfsdk:"vxlan"`
+	Facility           types.String `tfsdk:"facility"`
+	Metro              types.String `tfsdk:"metro"`
+	Description        types.String `tfsdk:"description"`
+	AssignedDevicesIds types.List   `tfsdk:"assigned_devices_ids"`
+}
+
+func (m *DataSourceModel) parse(vlan *packngo.VirtualNetwork) diag.Diagnostics {
+	m.ID = types.StringValue(vlan.ID)
+
+	if vlan.Project.ID != "" {
+		m.ProjectID = types.StringValue(vlan.Project.ID)
+	}
+
+	m.VlanID = types.StringValue(vlan.ID)
+	m.Description = types.StringValue(vlan.Description)
+	m.Vxlan = types.Int64Value(int64(vlan.VXLAN))
+
+	var metroCode, facilityCode types.String
+	if vlan.Facility != nil {
+		facilityCode = types.StringValue(vlan.Facility.Code)
+		metroCode = types.StringValue(strings.ToLower(vlan.Facility.Metro.Code))
+	}
+	// version of this resource. StateFunc doesn't exist in terraform and it requires implementation
+	// of bespoke logic before storing state. To ensure backward compatibility we ignore lower/upper
+	// case diff for now, but we may want to require input upper case
+	if !strings.EqualFold(m.Facility.ValueString(), facilityCode.ValueString()) {
+		m.Facility = facilityCode
+	}
+
+	if vlan.Metro != nil {
+		metroCode = types.StringValue(strings.ToLower(vlan.Metro.Code))
+	}
+	if !strings.EqualFold(m.Metro.ValueString(), metroCode.ValueString()) {
+		m.Metro = metroCode
+	}
+
+	deviceIds := make([]types.String, 0, len(vlan.Instances))
+	for _, device := range vlan.Instances {
+		deviceIds = append(deviceIds, types.StringValue(device.ID))
+	}
+
+	return m.AssignedDevicesIds.ElementsAs(context.Background(), &deviceIds, false)
+}
+
+type ResourceModel struct {
+	ID          types.String `tfsdk:"id"`
+	ProjectID   types.String `tfsdk:"project_id"`
+	Vxlan       types.Int64  `tfsdk:"vxlan"`
+	Facility    types.String `tfsdk:"facility"`
+	Metro       types.String `tfsdk:"metro"`
+	Description types.String `tfsdk:"description"`
+}
+
+func (m *ResourceModel) parse(vlan *packngo.VirtualNetwork) diag.Diagnostics {
+	m.ID = types.StringValue(vlan.ID)
+	m.Description = types.StringValue(vlan.Description)
+	m.Vxlan = types.Int64Value(int64(vlan.VXLAN))
+
+	if vlan.Project.ID != "" {
+		m.ProjectID = types.StringValue(vlan.Project.ID)
+	}
+
+	var metroCode types.String
+	facilityCode := types.StringNull()
+	if vlan.Facility != nil {
+		facilityCode = types.StringValue(vlan.Facility.Code)
```

Comment:
Rather than the EqualFold check below, this should be converted to lower-case just like metro code
```suggestion
		facilityCode = types.StringValue(strings.ToLower(vlan.Facility.Code))
```

Reply:
fixed